libs/utils/src/http/endpoint.rs

```diff
@@ -148,27 +147,164 @@ impl Drop for RequestCancelled {
+    // true or maybe get rid of the fully buffering path altogether
+    const CHUNKED_IS_DEFAULT: bool = false;
+
+    let chunked = super::request::parse_query_param::<_, bool>(&_req, "allow_chunked")?
+        .unwrap_or(CHUNKED_IS_DEFAULT);
+
+    let started_at = std::time::Instant::now();
+
+    if !chunked {
+        let span = info_span!("blocking");
+
+        tokio::task::spawn_blocking(move || {
+            let _span = span.entered();
+            // Currently we take a lot of mutexes while collecting metrics, so it's
+            // better to spawn a blocking task to avoid blocking the event loop.
+            let metrics = metrics::gather();
+            let mut buffer = Vec::new();
+            let encoder = TextEncoder::new();
+            encoder
+                .encode(&metrics, &mut buffer)
+                .map_err(|e| ApiError::InternalServerError(e.into()))?;
+
+            Ok(Response::builder()
+                .status(200)
+                .header(CONTENT_TYPE, encoder.format_type())
+                .body(Body::from(buffer))
+                .unwrap())
+        })
+        .await
+        .map_err(|e| ApiError::InternalServerError(e.into()))
+        .and_then(|e| e)
+    } else {
+        let (tx, rx) = mpsc::channel(1);
+
+        let body = Body::wrap_stream(ReceiverStream::new(rx));
+
+        struct ChannelWriter {
+            buffer: bytes::BytesMut,
+            tx: tokio::sync::mpsc::Sender<std::io::Result<bytes::Bytes>>,
+            written: usize,
+        }
 
-    let metrics = tokio::task::spawn_blocking(move || {
-        // Currently we take a lot of mutexes while collecting metrics, so it's
-        // better to spawn a blocking task to avoid blocking the event loop.
-        metrics::gather()
-    })
-    .await
-    .map_err(|e: JoinError| ApiError::InternalServerError(e.into()))?;
-    encoder.encode(&metrics, &mut buffer).unwrap();
+        impl ChannelWriter {
+            fn new(buf_len: usize, tx: mpsc::Sender<std::io::Result<Bytes>>) -> Self {
+                ChannelWriter {
+                    buffer: BytesMut::with_capacity(buf_len),
+                    tx,
+                    written: 0,
+                }
+            }
+
+            fn flush0(&mut self) -> std::io::Result<usize> {
+                let n = self.buffer.len();
+                if n > 0 {
+                    let ready = self.buffer.split().freeze();
+                    tracing::trace!(n = ready.len(), "flushing");
+
+                    // not ideal to call from blocking code to block_on, but we are sure that this
+                    // operation does not spawn_blocking other tasks
+                    let res: Result<(), ()> = tokio::runtime::Handle::current().block_on(async {
+                        self.tx.send(Ok(ready)).await.map_err(|_| ())?;
+
+                        // throttle sending to allow reuse of our buffer in `write`.
+                        self.tx.reserve().await.map_err(|_| ())?;
```

Comment:
This hard-depends on the channel having capacity 1.

I suggest to move channel creation into the `new()` function and return the writer and the stream as a tuple from there.

(That would also be a good opportunity to move this struct and impl into `utils`, it seems generic and useful).

Reply:
How does it hard depend on this?

> good opportunity to move

I don't want to expose this in any other place.

Reply:
Not a hard dependency. If you put (in the same function) a huge number as capacity, it might turn into like sparsely allocating all of the body depending on the slowloris client.